cg_openmm/simulation/physical_validation.py

```diff
@@ -0,0 +1,161 @@
+from simtk import unit
+import numpy as np
+from openmmtools.multistate import MultiStateReporter, MultiStateSampler, ReplicaExchangeSampler
+from openmmtools.multistate import ReplicaExchangeAnalyzer
+
+kB = (unit.MOLAR_GAS_CONSTANT_R).in_units_of(unit.kilojoule / (unit.kelvin * unit.mole))
+
+def physical_validation_ensemble(
+    temperature_list, ref_state_index=0, output_data="output.nc", output_directory="ouput", plotfile='ensemble_check'
+):
+    """
+    Run ensemble physical validation test for 2 states in replica exchange simulation
+
+    :param temperature_list: List of temperatures that will be used to define different replicas (thermodynamics states)
+    :type temperature_list: List( `SIMTK <https://simtk.org/>`_ `Unit() <http://docs.openmm.org/7.1.0/api-python/generated/simtk.unit.unit.Unit.html>`_ * number_replicas )
+
+    :param ref_state_index: Index in temperature_list to use as one of the states in the ensemble check. The other state will be chosen based on the energy standard deviation at the reference state
+    :type ref_state_index: int
+    
+    :param output_data: Path to the output data for a NetCDF-formatted file containing replica exchange simulation data
+    :type output_data: str
+    
+    :param plotfile: Filename for outputting ensemble check plot
+    :type plotfile: str
+    
+    """
+
+    # Read the energies for individual temperature replicas
+    reporter = MultiStateReporter(output_data, open_mode="r")
+    analyzer = ReplicaExchangeAnalyzer(reporter)
+
+    (
+        replica_energies,
+        unsampled_state_energies,
+        neighborhoods,
+        replica_state_indices,
+    ) = analyzer.read_energies()
+
+    n_particles = np.shape(reporter.read_sampler_states(iteration=0)[0].positions)[0]
+    temps = np.array([temp._value for temp in temperature_list])
+    beta_k = 1 / (kB * temps)
+    n_replicas = len(temperature_list)
+    for k in range(n_replicas):
+        replica_energies[:, k, :] *= beta_k[k] ** (-1)
+
+    total_steps = len(replica_energies[0][0])
+    state_energies = np.zeros([n_replicas, total_steps])
+
+    for step in range(total_steps):
+        for state in range(n_replicas):
+            state_energies[state, step] = replica_energies[
+                np.where(replica_state_indices[:, step] == state)[0], 0, step
+            ]
+
+    state_energies *= unit.kilojoule_per_mole
+
+
+    # Find optimal state pair for ensemble check:
+    # Compute standard deviations of each energy distribution:
+    state_energies_std = np.std(state_energies,axis=1)
+
+    # Select reference state point
+    T_ref = temperature_list[ref_state_index]
+    std_ref = state_energies_std[ref_state_index]
+
+    # Compute optimal spacing:
+    deltaT = 2*kB*T_ref**2/std_ref
+    #print("DeltaT: %r" %deltaT) 
+
+    # Find closest match
+    T_array = np.zeros(len(temperature_list))
+    for i in range(len(temperature_list)):
+        T_array[i] = temperature_list[i]._value

```

Comment:
Probably should be using value_in_unit in most cases?

Reply:
Ok I will update those - so value_in_unit just does an extra check to make sure it's in the units we want? We could also use that to avoid needing any conversions within physical validation (use value_in_unit with the Gromacs unit set)

Reply:
It actually gives the value in whatever units you specify as the argument.   _value gives the value in the units it is currently cast in, which is dangerous, since we generally don't know for sure what is passed in.